src/preprocess.rkt

```diff
@@ -1,44 +1,122 @@
 #lang racket
 
-(require "syntax/types.rkt" "points.rkt" "float.rkt")
-
-(provide preprocess-pcontext *herbie-preprocess* apply-preprocess)
-
-;; Tracks list of preprocess structs Herbie decides to apply
-(define *herbie-preprocess* (make-parameter empty))
-
-;; index-value-pairs is a sorted list of (index, value)
-(define (list-set-multiple list index-value-pairs)
-  (let loop ([current list] [todo index-value-pairs] [index 0])
-    (cond
-      [(empty? current)
-       empty]
-      [(and (not (empty? todo)) (equal? (first (first todo)) index))
-       (cons (second (first todo)) (loop (rest current) (rest todo) (+ index 1)))]
-      [else
-       (cons (first current) (loop (rest current) todo (+ index 1)))])))
-
-(define (apply-to-group variables point group-variables group-function)
-  (define indicies
-    (map (lambda (var) (index-of variables var)) group-variables))
-  (define values
-    (group-function (map (curry list-ref point) indicies)))
-  (define sorted (sort (map list indicies values) (lambda (a b) (< (first a) (first b)))))
-  (list-set-multiple point sorted))
-
-(define (sort-group variables point preprocess repr)
-  (match-define (list 'sort vars ...) preprocess)
-  (apply-to-group variables point vars (lambda (group) (sort group (curryr </total repr)))))
-
-(define (apply-preprocess sampled-point preprocess-structs ctx)
+(require "core/egg-herbie.rkt" "core/simplify.rkt"
+         "syntax/syntax.rkt" "syntax/types.rkt" "alternative.rkt" "common.rkt"
+         "programs.rkt" "points.rkt" "timeline.rkt" "float.rkt")
+
+(provide find-preprocessing preprocess-pcontext remove-unnecessary-preprocessing)
+
+;; See https://pavpanchekha.com/blog/symmetric-expressions.html
+(define (find-preprocessing expression context rules)
+  ;; Here `*` means a test identity that *may* be equal to `expression`, and
+  ;; `~` means the simplest form of an expression.
+  (define variables (context-vars context))
+  (define variable-representations (context-var-reprs context))
+  (define evens*
+    (for/list ([variable (in-list variables)]
+               [representation (in-list variable-representations)])
+      ;; TODO: Handle case where neg isn't supported for this representation
+      (define negate (get-parametric-operator 'neg representation))
+      (replace-vars (list (cons variable (list negate variable))) expression)))
+  (define pairs (combinations variables 2))
+  (define swaps*
+    (for/list ([pair (in-list pairs)])
+      (match-define (list a b) pair)
+      (replace-vars (list (cons a b) (cons b a)) expression)))
+  ;; TODO: If egraph detects unsoundness, abort preprocessing. Would need
+  ;; explicit access to egraph constructed by run-egg though
```

Comment:
I don't think this is necessary. If the egraph detects unsoundness, it automatically "rolls back" to the iter before that, which typically is OK.